Ensure torch.futures.wait_all exits early on error.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53953 Ensure torch.futures.wait_all exits early on error.**

torch.futures.wait_all, would wait for all specified futures to
complete before it returned. As a result, if there was an error it would still
wait for a long time (ex: long running RPCs) before it returned an error to the
user.

This PR ensures `wait_all` returns and error as soon as any future runs into an
error and doesn't wait for all futures to complete.

I removed the logic _invoke_rpc_python_udf which raised an error in the unwrap
function, because ideally the error should be set on the Future and not be
raised to the user only when `wait()` is called. As an example, in the case of
`wait_all`, the user never calls `wait()` on the future that errored out but a
future down the chain and we should propagate these errors via `setError`
instead.

#Closes: https://github.com/pytorch/pytorch/issues/46931

Differential Revision: [D27032362](https://our.internmc.facebook.com/intern/diff/D27032362/)

pritamdamania committed:
Ensure torch.futures.wait_all exits early on error.

torch.futures.wait_all, would wait for all specified futures to
complete before it returned. As a result, if there was an error it would still
wait for a long time (ex: long running RPCs) before it returned an error to the
user.

This PR ensures `wait_all` returns and error as soon as any future runs into an
error and doesn't wait for all futures to complete.

I removed the logic _invoke_rpc_python_udf which raised an error in the unwrap
function, because ideally the error should be set on the Future and not be
raised to the user only when `wait()` is called. As an example, in the case of
`wait_all`, the user never calls `wait()` on the future that errored out but a
future down the chain and we should propagate these errors via `setError`
instead.

Differential Revision: [D27032362](https://our.internmc.facebook.com/intern/diff/D27032362/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Ensure torch.futures.wait_all exits early on error."


torch.futures.wait_all, would wait for all specified futures to
complete before it returned. As a result, if there was an error it would still
wait for a long time (ex: long running RPCs) before it returned an error to the
user.

This PR ensures `wait_all` returns and error as soon as any future runs into an
error and doesn't wait for all futures to complete.

I removed the logic _invoke_rpc_python_udf which raised an error in the unwrap
function, because ideally the error should be set on the Future and not be
raised to the user only when `wait()` is called. As an example, in the case of
`wait_all`, the user never calls `wait()` on the future that errored out but a
future down the chain and we should propagate these errors via `setError`
instead.

#Closes: https://github.com/pytorch/pytorch/issues/46931

Differential Revision: [D27032362](https://our.internmc.facebook.com/intern/diff/D27032362/)

[ghstack-poisoned]

lw reviewed (commented):
Some food for thought:

If we compare the behavior of `collectAll` with other languages/libraries, there are some unusual differences: I think that typically such a function has a type signature `List[Future[T]] -> Future[List[T]]` (for example: [Python's asyncio](https://docs.python.org/3/library/asyncio-task.html#asyncio.gather), [C++'s Folly](https://github.com/facebook/folly/blob/master/folly/experimental/coro/Collect.h), [Hack](https://docs.hhvm.com/hack/reference/function/HH.Asio.v/)), whereas in this case here it has a signature of `List[Future[T]] -> Future[List[Future[T]]]`. This is not necessarily worse, and it even has some advantages. However it hints that `collectAll` will not "extract" anything from the input futures, it will just connect their signals and leave it to the user to then inspect their results. By setting the resulting future to error if any of the input ones fails, we're kinda shortcutting this abstraction.

Another interesting comparison is that many other languages/libraries end up cancelling/aborting other futures/coroutines when one of them fails. They do it for a few reasons: it avoids wasting resources to compute results that we already know will be discarded, and it also ensures that no future is left unawaited (which could lead to resource leaks). In PyTorch currently we don't have the "concept" of cancelling a future, but I think it might be interesting to add it. It could also for example help with timeouts.

pritamdamania87 commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -940,17 +940,24 @@ TORCH_API intrusive_ptr<ivalue::Future> collectAll(
   } else {
     auto typePtr = ctx->srcFutures.get(0)->elementType();
     for (const auto i : c10::irange(ctx->srcFutures.size())) {
-      TORCH_CHECK(i == 0 || *ctx->srcFutures.get(i)->elementType() == *typePtr);
```
I actually removed this on purpose since I felt the check doesn't make sense? The check is asserting that all the Futures contain the same type which seems to be too restrictive since you should be able to wait on a collection of futures irrespective of the underlying type they return. This was actually causing one of my tests to fail where one Future returned a PyObject and the other a Tensor.

lw replied:
Are you sure this will not create problems with TorchScript's type system? I though it wasn't able to handle non-homogeneous types, by which I mean that it requires a list to have type `List[T]`, hence have each item of the same type `T`. I don't know if in case of futures this requirement is relaxed, as it seems that future are a somewhat type-erased class... (in other words, in TorchScript's view, is this a `List[Future]` or a `List[Future[T]]`?)

If in your case you're expecting a mix of Tensors and PyObjects couldn't you solve it by using `Future[Union[Tensor, PyObject]]`?

pritamdamania87 replied:
My understanding is that TorchScript sees this as `List[Future]` and not `List[Future[T]]`.

> If in your case you're expecting a mix of Tensors and PyObjects couldn't you solve it by using Future[Union[Tensor, PyObject]]?

I'm not sure where exactly I would make this change? My unit test is completely in python and generates a PyObject for one rpc_async and a Tensor for the other.

lw replied:
> I'm not sure where exactly I would make this change? My unit test is completely in python and generates a PyObject for one rpc_async and a Tensor for the other.

Uhm, yeah, good point, the type of a future is determined by its creator, and we don't seem to have a way to change it or "cast it" to something else...

pritamdamania87 commented:
> If we compare the behavior of collectAll with other languages/libraries, there are some unusual differences: I think that typically such a function has a type signature List[Future[T]] -> Future[List[T]] (for example: Python's asyncio, C++'s Folly, Hack), whereas in this case here it has a signature of List[Future[T]] -> Future[List[Future[T]]].

I found this weird too when I was going through the code, although I'd like to keep the scope of this PR to fixing the `wait_all` early exit issue.

pritamdamania committed:
Update on "Ensure torch.futures.wait_all exits early on error."


torch.futures.wait_all, would wait for all specified futures to
complete before it returned. As a result, if there was an error it would still
wait for a long time (ex: long running RPCs) before it returned an error to the
user.

This PR ensures `wait_all` returns and error as soon as any future runs into an
error and doesn't wait for all futures to complete.

I removed the logic _invoke_rpc_python_udf which raised an error in the unwrap
function, because ideally the error should be set on the Future and not be
raised to the user only when `wait()` is called. As an example, in the case of
`wait_all`, the user never calls `wait()` on the future that errored out but a
future down the chain and we should propagate these errors via `setError`
instead.

#Closes: https://github.com/pytorch/pytorch/issues/46931

Differential Revision: [D27032362](https://our.internmc.facebook.com/intern/diff/D27032362/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Ensure torch.futures.wait_all exits early on error."


torch.futures.wait_all, would wait for all specified futures to
complete before it returned. As a result, if there was an error it would still
wait for a long time (ex: long running RPCs) before it returned an error to the
user.

This PR ensures `wait_all` returns and error as soon as any future runs into an
error and doesn't wait for all futures to complete.

I removed the logic _invoke_rpc_python_udf which raised an error in the unwrap
function, because ideally the error should be set on the Future and not be
raised to the user only when `wait()` is called. As an example, in the case of
`wait_all`, the user never calls `wait()` on the future that errored out but a
future down the chain and we should propagate these errors via `setError`
instead.

#Closes: https://github.com/pytorch/pytorch/issues/46931

Differential Revision: [D27032362](https://our.internmc.facebook.com/intern/diff/D27032362/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Ensure torch.futures.wait_all exits early on error."


torch.futures.wait_all, would wait for all specified futures to
complete before it returned. As a result, if there was an error it would still
wait for a long time (ex: long running RPCs) before it returned an error to the
user.

This PR ensures `wait_all` returns and error as soon as any future runs into an
error and doesn't wait for all futures to complete.

I removed the logic _invoke_rpc_python_udf which raised an error in the unwrap
function, because ideally the error should be set on the Future and not be
raised to the user only when `wait()` is called. As an example, in the case of
`wait_all`, the user never calls `wait()` on the future that errored out but a
future down the chain and we should propagate these errors via `setError`
instead.

#Closes: https://github.com/pytorch/pytorch/issues/46931

Differential Revision: [D27032362](https://our.internmc.facebook.com/intern/diff/D27032362/)

[ghstack-poisoned]

cbalioglu commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -940,17 +940,24 @@ TORCH_API intrusive_ptr<ivalue::Future> collectAll(
     for (const auto i : c10::irange(ctx->srcFutures.size())) {
-      TORCH_CHECK(i == 0 || *ctx->srcFutures.get(i)->elementType() == *typePtr);
+
+      auto fut = ctx->srcFutures.get(i);
```
I am here just to get a bit more context about RPC. My comment is mostly nitpick. Would it be more ideal to pass the index itself or the naked pointer (as the `intrusive_ptr` is already stored in `srcFutures` as part of lambda closure)? If not, you can also use the indexing operator instead of the range-checked `get()`.

pritamdamania87 replied:
Is the goal here to avoid the refcount bump for `intrusive_ptr`? It doesn't seem like that would be much of an issue given the use case here.

mrshenli commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -940,17 +940,24 @@ TORCH_API intrusive_ptr<ivalue::Future> collectAll(
+          return;
+        }
+
+        if (--ctx->remaining == 0 && !ctx->dstFuture->completed()) {
+          ctx->dstFuture->markCompleted(ctx->asIvalue);
```
can there be race in these two lines? `dstFuture->completed()` and `dstFuture->markCompleted()` are individually protected by a mutex, but what if two source Futures are processed by two different threads and calling into this concurrently? They can both think `completed()=false` and proceed to `markCompleted`?

pritamdamania87 replied:
I think that would be protected by `--ctx->remaining == 0`, that condition will be true for only one of the threads since remaining is atomic.

mrshenli replied:
Oh, I see, you are right!

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -3034,6 +3039,74 @@ def test_wait_all_raise_in_body(self):
+
+        dst = worker_name((self.rank + 1) % self.world_size)
+        fut1 = rpc.rpc_async(dst, RpcTest.timed_out_rpc)
+        fut2 = rpc.rpc_async(dst, raise_func_script, args=(expected_err,))
```
shall we add a test where multiple `rpc_async` can fail concurrently?

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -5,6 +5,7 @@
 import os
 import sys
 from threading import Lock
+from threading import Event
```
Sort imports?

mrshenli reviewed (commented):
LGTM! Left one question on whether we should worry about a potential raise.

pritamdamania committed:
Update on "Ensure torch.futures.wait_all exits early on error."


torch.futures.wait_all, would wait for all specified futures to
complete before it returned. As a result, if there was an error it would still
wait for a long time (ex: long running RPCs) before it returned an error to the
user.

This PR ensures `wait_all` returns and error as soon as any future runs into an
error and doesn't wait for all futures to complete.

I removed the logic _invoke_rpc_python_udf which raised an error in the unwrap
function, because ideally the error should be set on the Future and not be
raised to the user only when `wait()` is called. As an example, in the case of
`wait_all`, the user never calls `wait()` on the future that errored out but a
future down the chain and we should propagate these errors via `setError`
instead.

#Closes: https://github.com/pytorch/pytorch/issues/46931

Differential Revision: [D27032362](https://our.internmc.facebook.com/intern/diff/D27032362/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM!
